src/app/users/get-user.ts

```diff
@@ -1,65 +1,99 @@
-import {
-  SemanticAttributes,
-  asyncRunInSpan,
-  addAttributesToCurrentSpan,
-  ENDUSER_ALIAS,
-} from "@services/tracing"
+import { SemanticAttributes, asyncRunInSpan } from "@services/tracing"
 import { UsersRepository } from "@services/mongoose"
-import { IpFetcher } from "@services/ipfetcher"
 import { getIpConfig } from "@config/app"
+import { IpFetcher } from "@services/ipfetcher"
+import { UsersIpRepository } from "@services/mongoose/users-ip"
 import { RepositoryError } from "@domain/errors"
 
 const users = UsersRepository()
+const usersIp = UsersIpRepository()
 
 export const getUserForLogin = async ({
   userId,
   ip,
+  logger,
 }: {
-  userId: string
-  ip?: string
+  userId: UserId
+  ip?: Ip
+  logger: Logger
 }): Promise<User | ApplicationError> =>
   asyncRunInSpan(
     "app.getUserForLogin",
     { [SemanticAttributes.CODE_FUNCTION]: "getUserForLogin" },
     async () => {
-      const user = await users.findById(userId as UserId)
+      const user = await users.findById(userId)
+
       if (user instanceof Error) {
         return user
       }
-      addAttributesToCurrentSpan({
-        [ENDUSER_ALIAS]: user.username,
-      })
-      user.lastConnection = new Date()
 
-      // IP tracking logic could be extracted into domain
+      // this routing run asynchrously, to update metadata on the background
+      updateUserIpInfo({ userId, ip, logger })
+
+      return user
+    },
+  )
+
+const updateUserIpInfo = async ({
+  userId,
+  ip,
+  logger,
+}: {
+  userId: UserId
+  ip?: Ip
+  logger: Logger
+}): Promise<void> =>
+  asyncRunInSpan(
+    "app.addIp",
+    { [SemanticAttributes.CODE_FUNCTION]: "addIp" },
+    async () => {
       const ipConfig = getIpConfig()
-      if (ip && ipConfig.ipRecordingEnabled) {
-        const lastIP: IPType | undefined = user.lastIPs.find(
-          (ipObject: IPType) => ipObject.ip === ip,
-        )
-        if (lastIP) {
-          lastIP.lastConnection = user.lastConnection
-        } else if (ipConfig.proxyCheckingEnabled) {
+
+      const lastConnection = new Date()
+
+      const userIp = await usersIp.findById(userId)
+
+      if (userIp instanceof RepositoryError) throw userIp
```

Comment:
this should be return here not throw right? @bodymindarts @dolcalmi ?

Reply:
Correct in app layer always return the error